core-api/src/main/java/com/optimizely/ab/Optimizely.java

```diff
@@ -1126,6 +1129,200 @@ public OptimizelyConfig getOptimizelyConfig() {
         return new OptimizelyConfigService(projectConfig).getConfig();
     }
 
+    //============ decide ============//
+
+    /**
+     * Create a context of the user for which decision APIs will be called.
+     *
+     * A user context will be created successfully even when the SDK is not fully configured yet.
+     *
+     * @param userId The user ID to be used for bucketing.
+     * @param attributes: A map of attribute names to current user attribute values.
+     * @return An OptimizelyUserContext associated with this OptimizelyClient.
+      */
+    public OptimizelyUserContext createUserContext(@Nonnull String userId,
+                                                   @Nonnull Map<String, Object> attributes) {
+        if (userId == null) {
+            logger.warn("The userId parameter must be nonnull.");
+            return null;
+        }
+
+        return new OptimizelyUserContext(this, userId, attributes);
+    }
+
+    public OptimizelyUserContext createUserContext(@Nonnull String userId) {
+        return new OptimizelyUserContext(this, userId);
+    }
+
+    OptimizelyDecision decide(@Nonnull OptimizelyUserContext user,
+                              @Nonnull String key,
+                              @Nonnull List<OptimizelyDecideOption> options) {
+
+        ProjectConfig projectConfig = getProjectConfig();
+        if (projectConfig == null) {
+            return OptimizelyDecision.newErrorDecision(key, user, DecisionMessage.SDK_NOT_READY.reason());
+        }
+
+        FeatureFlag flag = projectConfig.getFeatureKeyMapping().get(key);
+        if (flag == null) {
+            return OptimizelyDecision.newErrorDecision(key, user, DecisionMessage.FLAG_KEY_INVALID.reason(key));
+        }
+
+        String userId = user.getUserId();
+        Map<String, Object> attributes = user.getAttributes();
+        Boolean decisionEventDispatched = false;
+        List<OptimizelyDecideOption> allOptions = getAllOptions(options);
+        DecisionReasons decisionReasons = DefaultDecisionReasons.newInstance(allOptions);
+
+        Map<String, ?> copiedAttributes = new HashMap<>(attributes);
+        FeatureDecision flagDecision = decisionService.getVariationForFeature(
+            flag,
+            userId,
+            copiedAttributes,
+            projectConfig,
+            allOptions,
+            decisionReasons);
```

Comment:
It appears that `decisionReasons` is expected to be modified from inside. In my experience, this kind of flow gets a little confusing when someone is following the code flow. On suggestion can be to return `Pair<Decision, Reasons>` from`getVariationForFeature`. This will clearly tell that decision reasons are also coming back from this function and original object will also be intact causing no confusion. what do you think

Reply:
A good point! We also consider that option as well. It's dropped since it's a bit complicated to process and merge (and/or audience, etc) at each stage. We can find better solutions. I think you can try with other sdks with your suggestion and I can come back to this if we see a good solution.